Supporting multiple network interfaces

PetrichorIT opened the pull request:
This PR changes:
- adds `NodeIdentifiers` to decouple node addresses and node identities 
- adds `IpSubnets` to model different subnets (subnets are not yet supported, see #132 for progress)
- adds `1:n` mappings between names and `IpAddr` in `dns.rs`
- decouple dns registration from dns lookups (lookups of nonexistent names will no longer return a newly allocated address)
- integrales reverse dns lookups into `ToIpAddr` trait 


relates to  #132

mcches commented:
Apologies for the delay in review. I have this queued up for tomorrow morning (08/15).

mcches commented on src/ip.rs:
```diff
@@ -1,4 +1,283 @@
+pub(crate) struct ScopedIpAddr {
+    pub(crate) addr: IpAddr,
+    #[allow(dead_code)]
+    pub(crate) subnet: IpSubnet,
```
I'm struggling a bit with the subnet part. This factoring makes it feel like the host _has a subnet_, since the host has one or more of these.

I wonder if we should get the network interfaces right first and then add subnet support.

mcches reviewed (commented):
Thanks for getting started on this.

PetrichorIT commented on src/dns.rs:
```diff
@@ -12,19 +15,27 @@ use crate::ip::IpVersionAddrIter;
+    fn to_ip_addr(&self, dns: &Dns) -> Option<IpAddr>;
+
+    #[doc(hidden)]
+    fn to_name(&self, dns: &Dns) -> String;
```
This part is only temporarily necessary.

The problem is, that we need a name to create a dns mapping. However not all nodes have a name using the 
current API since `Sim::client` expects a `impl ToIpAddr`. So to get a name for the dns registration,
this trait must temporarily be able to create such a name.

In the future all nodes should have explicit names, independent of their bound addresses, so this will only be temporary. Explicit names could be archived with the proposed builder API for node creation.

mcches replied:
I see, so you want to avoid a breaking change in this first change? If so, can you comment that here saying we'll remove this.

I'm inclined to just break things and work in a remote branch if that allows for less churn.

PetrichorIT commented on src/dns.rs:
```diff
@@ -37,79 +48,139 @@ impl Dns {
+        let scoped_addr = addr
+            .to_ip_addr(self)
+            .map(|addr| {
+                if addr.is_ipv4() {
```
like this?
```rust
let scoped_addr = match addr.to_ip_addr(self).map(|addr| addr) {
    Some(addr) if addr.is_ipv4() => ScopedIpAddr {
        addr,
        subnet: IpSubnet::V4(Ipv4Subnet::default()),
    },
    Some(addr) /*if addr.is_ipv6() */=> ScopedIpAddr {
        addr,
        subnet: IpSubnet::V6(Ipv6Subnet::default()),
    },
    None => self.addrs.next(),
};
```

PetrichorIT commented on src/node.rs:
```diff
@@ -0,0 +1,21 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct NodeIdentifer {
+    node_name: Arc<str>,
```
The advantage of using str is the fact, that a `NodeIdentifier` not only acts as a unique identifier, but can also be used to create tracing spans. So for the time being, I think we should stick to strs.

mcches replied:
You can have both, right? A uuid does all the network mapping and the host that is active in the sim loop still has a human readable hostname that is set on the span before each tick.

PetrichorIT commented:
So in the spirit of breaking things, this PR fully supports multiple IP subnets, thus multiple bound addresses.

# Changes

- Nodes are identified by a `NodeIdentifier`, thus they always have a human-readable name
- All nodes in the simulation are now connected through a list of subnets provided to `Builder::ip_subnets`
    - The default subnets are (IPv4 192.168.0.0/16 + IPv6 fe80::/64)
    - Each node is assigned an address in every subnet. Addresses can be specified with `NodeBuilder::with_addr` or are automatically assigned in the node creation process
- Node creation can now be managed using the `NodeBuilder` API
    - `Sim::client` and `Sim::host` are still supported, but may be removed
    - `Sim::client` and `Sim::host` no longer support `impl ToIpAddr` as  input, only `impl AsRef<str>` is supported, since every node needs are name (provided input is only used as a name, not longer as an addr)
- DNS changes
    - `lookup` will no longer be used to register nodes (consider it read-only from now on)
    - `Dns::register` can be used to register new nodes in the DNS resolution
    - since every node has a name, every node is represented in the DNS struct
    - Reverse lookups now return a node identifier, instead of a string
    - removed `trait ToIpAddr`, since all queries may return multiple IP address
    - the public API was already using `trait ToIpAddrs`, so no changes there
    - internally `for_pairs` was replaced by `for_connected_pairs`, since only IPs in the same subnet are connected
    - `trait ToSocketAddrs` can now return multiple addrs
- Topology
    - Only IPs in the same subnet are connected in the topology
    - `endpoints` keeps track of all addresses in one subnet, for easier node creation (maybe broadcast in the future?)
- Host specific
    - Hosts are now bound to multiple addrs
    - `net::tcp` and `net::udp` bind / connect to appropriate addresses 
- removed `IpVersion` and `Builder::ip_version`
- removed `*_many` APIs

# Open questions

### Builder API

The node builder API is somewhat cumbersome, when creating very simple nodes. Notably `build_client` or `build_host`
seems very verbose.
```rust
sim.node("my-simple-node").build_client(async { ... });
```
We could implement the API in such a way to allow these statements,
```rust
sim.client("my-simple-node").build(async { ... });
sim.host("my-simple-host").build(async { ... });
```
but that would use up the fn names `Sim::client` and `Sim::host`. Alternatively
we could keep the original `Sim::client` / `Sim::host` as simple aliases to the fully qualified form.

mcches commented:
I've not gotten the time to review this, many applogies. Another issue just came up that can provide further motivation for prioritizing this work.

See #153

progwriter commented:
Hi folks, any update here? 

This looks like a super useful feature for modeling systems in networking, where multi-interfaces are a must. Would be great to see this available in turmoil!

mcches commented:
> Hi folks, any update here?
> 
> This looks like a super useful feature for modeling systems in networking, where multi-interfaces are a must. Would be great to see this available in turmoil!

Thanks for checking in. This is low priority at the moment. I've got a decent refactor underway for the net modules and after that is in I can pick up contributing to this or guiding the changes.

Can you please elaborate on your use cases/requirements so that we take them into consideration?

progwriter commented:
> Can you please elaborate on your use cases/requirements so that we take them into consideration?

At a high level, doing some testing of a network routing protocol. Since the protocol uses a mix of tcp + udp, was hoping to use turmoil to remove networking non-determinism and enable contained testing + fault injection.

Being able to treat routers as having multiple interfaces is pretty important (for link-local and loopback-to-loopback communication)

mcches commented:
Closing due to inactivity and because we need to refactor some core things for this to work.
